File: Week1/2p_games/Greedy Or Not/Greedy or not.py
print("Enter the size of the list")
n = int(input())
L = list(map(int, 
    input("Enter the numbers :\n").strip().split()))
print(L)
temp = L.copy()
for length in range(2,n+1):
    temp = [ max(L[start]-temp[start+1],L[start+length-1]-temp[start]) for start in range(0,n-length+1) ]
    
if temp[0] > 0:
    print("Player1 wins")
elif temp[0] == 0:
    print("Draw")
else:
    print("Player2 wins")

File: Week1/2p_games/Greedy Or Not/Greedy or not.cpp
#include<iostream>
using namespace std;
int main()
{
    cout << "Enter the size of the list\n";
    int n;
    cin>>n;
    int a[n];
    cout << "Enter the elements of the list\n";
    for (int i = 0 ; i < n ; i++)
    {
        cin >> a[i];
    }

    // dp[i][j] represents the difference between points of first and second player if the indices between i and i+j-1 (including both) are presented(i.e. numbers in those indices are the only ones valid for playing) and both play optimally.
    int dp[n][n+1];
    for (int start = 0 ; start < n ;start++)
    {
        dp[start][1] = a[start];
    }
    for (int len = 2 ; len<=n ; len++)
    {
        for (int start = 0 ; start <= n-len ; start++)
        {
            dp[start][len]= max(a[start]-dp[start+1][len-1],a[start+len-1]-dp[start][len-1]);
        }
    }
    if (dp[0][n] > 0)
    {
        cout << "Player1 wins";
    } 
    else if (dp[0][n] == 0)
    {
        cout << "Draw";
    }
    else
    {
        cout << "Player2 wins";
    }
}
